pulsar-io/kinesis/src/main/java/org/apache/pulsar/io/kinesis/KinesisRecordProcessor.java

```diff
@@ -0,0 +1,108 @@
+
+import java.nio.charset.CharacterCodingException;
+import java.nio.charset.Charset;
+import java.nio.charset.CharsetDecoder;
+import java.util.List;
+import java.util.concurrent.LinkedBlockingQueue;
+
+import com.amazonaws.services.kinesis.clientlibrary.exceptions.InvalidStateException;
+import com.amazonaws.services.kinesis.clientlibrary.exceptions.KinesisClientLibDependencyException;
+import com.amazonaws.services.kinesis.clientlibrary.exceptions.ShutdownException;
+import com.amazonaws.services.kinesis.clientlibrary.exceptions.ThrottlingException;
+import com.amazonaws.services.kinesis.clientlibrary.interfaces.IRecordProcessor;
+import com.amazonaws.services.kinesis.clientlibrary.interfaces.IRecordProcessorCheckpointer;
+import com.amazonaws.services.kinesis.clientlibrary.lib.worker.ShutdownReason;
+import com.amazonaws.services.kinesis.model.Record;
+
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class KinesisRecordProcessor implements IRecordProcessor {
+    
+    private final int numRetries;
+    private final long checkpointInterval;
+    private final long backoffTime;
+    
+    private final LinkedBlockingQueue<KinesisRecord> queue;
+    private long nextCheckpointTimeInMillis;
+    private String kinesisShardId;
+    
+    public KinesisRecordProcessor(LinkedBlockingQueue<KinesisRecord> queue, KinesisSourceConfig config) {
+        this.queue = queue;
+        this.backoffTime = config.getBackoffTime();
+        this.checkpointInterval = config.getCheckpointInterval();
+        this.numRetries = config.getNumRetries();
+    }
+
+    @Override
+    public void initialize(String shardId) {
+        kinesisShardId = shardId;
+    }
+
+    @Override
+    public void processRecords(List<Record> records, IRecordProcessorCheckpointer checkpointer) {
+        log.info("Processing " + records.size() + " records from " + kinesisShardId);
+        
+        for (Record record : records) {
+           try {
+               queue.put(new KinesisRecord(record));
+           } catch (InterruptedException e) {
+               log.error("unable to create KinesisRecord ", e);
+           }
+        }
+
+       // Checkpoint once every checkpoint interval.
+        if (System.currentTimeMillis() > nextCheckpointTimeInMillis) {
+            checkpoint(checkpointer);
+            nextCheckpointTimeInMillis = System.currentTimeMillis() + checkpointInterval;
+        }
+    }
+
+    @Override
+    public void shutdown(IRecordProcessorCheckpointer checkpointer, ShutdownReason reason) {
+        log.info("Shutting down record processor for shard: " + kinesisShardId);
+        if (reason == ShutdownReason.TERMINATE) {
+            checkpoint(checkpointer);
+        }
+    }
+    
+    private void checkpoint(IRecordProcessorCheckpointer checkpointer) {
+        log.info("Checkpointing shard " + kinesisShardId);
+        
+        for (int i = 0; i < numRetries; i++) {
+
+            try {
+                checkpointer.checkpoint();
```

Comment:
This loop never breaks on success. It should break out if the checkpoint is successful, rather than running the operation multiple times.